### PR TITLE
Core/3.0

### PR DESCRIPTION
**Pull request by mrkno**

### CODE REVIEW

File: core/config.js [hunk 1]
```diff
@@ -1,42 +1,121 @@
...
+ *		Copyright (c) Matthew Knox and Contributors 2015.
+ */
 
-exports.getConfig = function(m) {
-	if (loading) {
-		throw 'Cannot get config while loading.';
-	}
+var fs              = require('fs'),
+    path            = require('path'),
+	modes			= require('./modes.js').listModes(),
```
Comment (Awarua-) [inline core/config.js]: indentation

File: core/config.js [hunk 2]
```diff
@@ -1,42 +1,121 @@
...
+    var isSystem = sysConfigZones.includes(m);
+    
+    if (!sysConfig) {
+        sysConfig = loadConfig(sysConfigFile);
+    }
+    
+    if (!isSystem && !sysConfig[m]) {
+        return {};
+    }
+	else if (!sysConfig[m]) {
```
Comment (Awarua-) [inline core/config.js]: Identation

File: core/config.js [hunk 3]
```diff
@@ -1,42 +1,121 @@
...
+    }
+
+    var loc = path.join(location, modConfigFile),
+        configData = loadConfig(loc),
+        deprecatedDataA = exports.getConfig(module.name),
+        deprecatedDataB = exports.getConfig(module.name + '.js');
+    for (var name in deprecatedDataA) {
+        configData[name] = deprecatedDataA[name];
+    }
+	for (var name in deprecatedDataB) {
```
Comment (Awarua-) [inline core/config.js]: Identation

File: core/core_modules/uptime.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+var startTime;
+
+exports.match = function(text, commandPrefix) {
+  return text === commandPrefix + 'uptime';
+};
+
+exports.run = function(api, event) {
+  var date = new Date(),
+    seconds = (date.getTime() - startTime.getTime()) / 1000;
+  api.sendMessage('I\'ve been alive for ' + seconds + ' seconds.', event.thread_id);
```
Comment (Awarua-) [inline core/core_modules/uptime.js]: Could this be improved like Months days minutes seconds

File: core/modules.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            if (stat == null) {
+                continue;
+            }
+
+            var kj = require.once(p);            
+            if (!kj.name || !kj.startup) {
+                continue;
+            }
+            
+			if (exports.disabledConfig && exports.disabledConfig[kj.name] && exports.disabledConfig[kj.name] === true) {
```
Comment (Awarua-) [inline core/modules.js]: Indentation

File: core/modules.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        try {
+            if (index !== -1) {
+                console.write("Reloading module: '" + module.name + "'... " + (console.isDebug() ? "\n" : ""));
+                m = require.reload(startPath);
+            } else {
+                console.write("Loading module '" + module.name + "'... " + (console.isDebug() ? "\n" : ""));
+                m = require.once(startPath);
+            }
+        } catch (e) {
+            throw 'Could not require module \'' + module.name + '\'. Does it have a syntax error?';
```
Comment (Awarua-) [inline core/modules.js]: Would "load" be a better description than require?

File: core/modules.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        }
+        m.config = config.loadModuleConfig(module, modulePath);
+		m.name = module.name;
```
Comment (Awarua-) [inline core/modules.js]: Indentation

File: core/platform.js [hunk 8]
```diff
@@ -1,232 +1,208 @@
...
-        delete require.cache[mod.id];
-    });
+var figlet = require('figlet'),
+ 
+Platform = function(modes) {
+	require.reload('./prototypes.js');
+	
+	this.config			= require('./config.js');
+	this.loadedModules	= [];
+	this.coreModules	= [];
```
Comment (Awarua-) [inline core/platform.js]: tabbing / Indentation

File: core/platform.js [hunk 9]
```diff
@@ -1,232 +1,208 @@
...
+		if (this.loadedModules[i].match.apply(this.loadedModules[i], matchArgs)) {
 			try {
-				loadedModules[i].run(api, event);
+				this.handleTransaction(this.loadedModules[i], runArgs);
 			}
-			catch(e) {
-				api.sendMessage(event.body + ' fucked up. Damn you ' + event.sender_name + ".", event.thread_id);
-				console.trace(e);
+			catch (e) {
+                api.sendMessage(event.body + ' fucked up. Damn you ' + event.sender_name + ".", event.thread_id);
```
Comment (Awarua-) [inline core/platform.js]: Might want to remove the profanity

File: core/platform.js [hunk 10]
```diff
@@ -1,232 +1,208 @@
...
-			started = true;
-			console.log('System has started. Hello World!');
-		});
-	});
-};
+	if (!this.modes.length) {
+		throw 'Modes must be set before starting';
+    }
+
+	console.title(figlet.textSync(this.packageInfo.name.toProperCase()));
```
Comment (Awarua-) [inline core/platform.js]: Indentation

File: core/platform.js [hunk 11]
```diff
@@ -1,232 +1,208 @@
...
+        this.coreModules[i] = null;
+    }
+    this.coreModules = [];
+
+	this.config.saveConfig();
+	this.statusFlag = flag ? flag : StatusFlag.Shutdown;
+	
+	console.warn(this.packageInfo.name + " has shutdown.");
+	if (this.onShutdown && this.onShutdown != null) {
+		this.onShutdown(this.statusFlag);
```
Comment (Awarua-) [inline core/platform.js]: Should this be this.setOnShutdown(this.statusFlag); ??

File: core/startup.js [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+ * License:
+ *		MIT License. All code unless otherwise specified is
+ *		Copyright (c) Matthew Knox and Contributors 2015.
+ */
+
+var selectedModes		= null,
+	startNewPlatform	= function() {
+		try {
+			var Platform = require.once('./platform.js'),
+				platform = new Platform(selectedModes);
```
Comment (Awarua-) [inline core/startup.js]: passing null into Platform here ??

File: core/unsafe/console.js [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+ *		Copyright (c) Matthew Knox and Contributors 2015.
+ */
+
+var colours = require.safe('colors'),
+	fs		= require('fs'),
+	strip	= require.safe('stripcolorcodes'),
+    info    = console.info,
+    error   = console.error,
+    warn    = console.warn,
+	write	= process.stdout.write,
```
Comment (Awarua-) [inline core/unsafe/console.js]: All the weird indentation

File: main.js [hunk 14]
```diff
@@ -11,97 +12,66 @@
...
 
 // Determine if debug output is enabled
-var debug = false;
-if (process.argv[2] === 'debug') {
-	process.argv.splice(2, 1);
-	debug = true;
-	platform.debug = true;
+if (arguments[0] === 'debug') {
+    console.warn('Debug mode enabled.');
+	arguments.splice(0, 1);
```
Comment (Awarua-) [inline main.js]: Indentation

**Comment by Awarua-** [comment]
Looking good, It would be nice if there was a developer readme, to document all the new things that have been added, and things that people should be aware of

File: core/platform.js (hunk 9 above)
Comment (mrkno) [inline core/platform.js]: It has been there for a very long time, I don't think it needs to change

File: core/platform.js (hunk 11 above)
Comment (mrkno) [inline core/platform.js]: no

File: core/startup.js (hunk 12 above)
Comment (mrkno) [inline core/startup.js]: If you read the logic, it will never get to that point

**Pull request merged**